### ISSUE TITLE
lints to disable in tests

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust-clippy#12581]
### Description

In [Arti](https://gitlab.torproject.org/tpo/core/arti) have things like the following:
```
#[cfg(test)]
mod test {
    // @@ begin test lint list maintained by maint/add_warning @@
    #![allow(clippy::bool_assert_comparison)]
    #![allow(clippy::clone_on_copy)]
    #![allow(clippy::dbg_macro)]
    #![allow(clippy::mixed_attributes_style)]
    #![allow(clippy::print_stderr)]
    #![allow(clippy::print_stdout)]
    #![allow(clippy::single_char_pattern)]
    #![allow(clippy::unwrap_used)]
    #![allow(clippy::unchecked_duration_subtraction)]
    #![allow(clippy::useless_vec)]
    #![allow(clippy::needless_pass_by_value)]
    //! <!-- @@ end test lint list maintained by maint/add_warning @@ -->
    // TODO add this next lint to maint/add_warning, for all tests
    #![allow(clippy::iter_overeager_cloned)]
```

Our `maint/add_warning` is a Python script that keeps the (currently) 236 copies of this lint block in sync across our 61 crates.

It would be nice if there were a way to systematically and centrally disable these lints for all our tests.  Probably, many of them should be disabled by default.

(@llogiq suggested I should file this ticket)

**Comment by llogiq** [comment]
Thank you!

**Comment by llogiq** [comment]
Finally I got around to looking into your list. I found that most lints will already ignore test code if configured so:

* `dbg_macro` has an `[allow-dbg-in-tests](https://doc.rust-lang.org/nightly/clippy/lint_configuration.html#allow-dbg-in-tests)` configuration. If you set that to `true` in your `clippy.toml`, you can remove that lint from your list globally.
* [allow-print-in-tests](https://doc.rust-lang.org/nightly/clippy/lint_configuration.html#allow-print-in-tests)`=true` takes care of both `print_stdout` and `print_stderr`.
* `single_char_pattern` has been demoted to `pedantic` anyway, so it won't be linted at all unless you explicitly activate it.
* [allow-unwrap-in-tests](https://doc.rust-lang.org/nightly/clippy/lint_configuration.html#allow-unwrap-in-tests) makes `unwrap_used` ignore tests
* I just pushed #12725 to implement the same configuration option for `useless_vec`

I'll have a look at the other lints when I find a few spare minutes.

**Comment by ethus3h** [comment]
Yay, glad to find `allow-unwrap-in-tests`! Thanks.

For what it's worth, I've similarly been disabling these others for my tests, and it would be nice to have allow-...-in-tests options for them: `#[allow(clippy::unwrap_in_result, clippy::panic_in_result_fn)]`

**Comment by jultty** [comment]
An option for allowing `panic_in_result_fn` in tests would be particularly useful since, otherwise, for anything not already returning a matching propagation-ready error, you have to write some silly code that looks like an assert knockoff:

```rs
if !error_message.contains("not valid unicode") {
    return Err(Error::from(
        "Different error where UTF8 error was expected"
    ))
}

if template_count != default_count {
    return Err(Error::from(
        "Expected template number equal to default"
    ))
}
```

**Comment by mkatychev** [comment]
Is there a proposal for a generic way to specify a lint should not be included in tests? It seems like a lot of these implementations of `allow-*-in-tests` should be included in a list.
